### ISSUE TITLE
Feature: Use NestJS integration as SSR/http server

### ISSUE THREAD

**Issue by brandonroberts** [analogjs/analog#317]
### Which scope/s are relevant/related to the feature request?

platform

### Information

NestJS is very popular Node.js framework for building APIs and is widely used in the Angular ecosystem. This feature would allow NestJS to be used as the server for the `/api` routes, and provided integration as the backend for SSR.

Links:
- https://www.npmjs.com/package/vite-plugin-node
- https://docs.nestjs.com/

In development, NestJS would be configured under the `/api` path. During the build it would host the Analog application and provide additional configured API routes.

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [ ] No

**Comment by Yberion** [comment]
Hello, which adapter will use the implementation?

It would be great if it could use the Fastify adapter by default. 😄

**Comment by brandonroberts** [comment]
It will be set up in a way where you can control the implementation. There will be an API "entry point" where you can add additional configuration to the NestJS/Fastify adapter.

**Comment by goetzrobin** [comment]
Seems like this could be helpful:
https://github.com/wjq990112/nestjs-ssr

**Comment by nhvu95** [comment]
Does anyone currently work on this?

**Comment by goetzrobin** [comment]
@brandonroberts how should routing work with nest? Since nest has the concept of controllers which themselves have a path and methods inside of these controllers that can have their own path, I think it might be better to just require an app module inside the server folder and then let people define their own structure inside that folder however they please?

**Comment by brandonroberts** [comment]
Yea, that's where I'm leaning also. Integrating with Nest would be more manual in some regards, because there's no file based routing, and developers would have more control over the server integration

**Comment by brandonroberts** [comment]
@nhvu95 no one is working on this yet

**Comment by nhvu95** [comment]
At the moment, I created a separate NestJS App; then I made it work by using `http-proxy-middleware` to forward requests from Vite to NestJS.

```javascript
import { createProxyMiddleware } from 'http-proxy-middleware';

viteServer.middlewares.use(apiPrefix,  createProxyMiddleware({
          target: 'http://localhost:7749',
          changeOrigin: true,
          pathRewrite: {
            '^/api': '',
          }}
        ) as any);
```

I still think of a better way, like making them work in the same Vite port like Nitro did. In the link @goetzrobin post. They push the  Vite into the NestJS app, and now I looking for a way to do the opposite, push Nest into Vite.

**Comment by goetzrobin** [comment]
@brandonroberts @nhvu95 I can spend some time on this if you are open to that

**Comment by goetzrobin** [comment]
Haven't made too much progress unfortunately. I am running into issues with loadSSRModule trying to use the vite-plugin-node and more.
I did find this project however: https://github.com/cyco130/vavite and it has an example that might be helpful.
Specifically these examples: 
- https://github.com/cyco130/vavite/tree/main/examples/nestjs
- https://github.com/cyco130/vavite/tree/main/examples/nestjs-vite-plugin-ssr

@brandonroberts just to confirm my overall understanding of what is going on is correct:
In DEV: vite dev server renders our Angular application (and ideally runs the NestJs Express or Fastify server as a middleware under the /api route)
In PROD: vite builds static assets to the public assets folder of the NestJs Express or Fastify server and our code makes that server aware how to render Angular on the server.

Does that sounds about right?

**Comment by brandonroberts** [comment]
Yes, that sounds right. Maybe we could start by providing an example repo the two working together as a starting point. No rush on this though as we already have a working server solution in Analog and the trpc integration

**Comment by marcus-sa** [comment]
https://deepkit.io would be even better.
They have a RPC framework https://deepkit.io/library/rpc

**Comment by Flusinerd** [comment]
Okay, after talking to @brandonroberts I managed to get it working: https://github.com/Flusinerd/analog-nestjs-example

It's a mess atm and I need to clean it up.

Basically it comes down to this part:
```
server.middlewares.use('/api', <vite-plugin-node implementation code>)
```` 

Unfortunately there is no way to customise this behaviour in the ``vite-plugin-node`` implementation, so we either need to fork it or ask the author, if he is willing to accept a PR to implement this customization.

**Comment by Flusinerd** [comment]
For the integration with NestJS we wanna strive for a gRPC integration using some sort of proxy or server-services:

https://github.com/mgechev/ts-rpc
https://github.com/PatrickJS/angular-server-services
https://github.com/kmturley/angular-nest-grpc

**Comment by jeremylcarter** [comment]
I am happy to help out on this issue. Who is coordinating the effort?

**Comment by Flusinerd** [comment]
Hi @jeremylcarter .
I have done some work but  ran into a roadblock. If you want you could take a look at this:
https://github.com/Flusinerd/analog-nestjs-example/tree/ssr

Basically the issue is that ``import.meta.glob`` does not exist, when using rollup and therefore the analog router can't be imported. If you remove the import of the router in ``src/app/app.config.ts``, then everything works and everything works, except for the router ofc.

If you have further questions, I'd be happy to help. I also had a bit of a conversation about this on the analog discord inside the Contributing -> ideas channel.

**Comment by jeremylcarter** [comment]
Thanks! I'll have a look

**Comment by cskiwi** [comment]
@jeremylcarter Any update/progress on the nestjs integration?

**Comment by brandonroberts** [comment]
Made some progress here with efforts from @cskiwi 

Able to serve/build the nestjs server along with the Analog app in this repo

https://github.com/brandonroberts/nestjs-analog

SSR support still needs work

**Comment by cskiwi** [comment]
@brandonroberts oh nice. I'll try and find some time to look at the SSR, unless you figured it out by then :)

**Comment by cskiwi** [comment]
@brandonroberts, been a bit busy. Any changes regarding the NestJS "integration".

**Comment by brandonroberts** [comment]
Same here. No new movement on it as of late

**Comment by jeremylcarter** [comment]
Sorry everyone. I was planning on looking at this integration, but I got distracted by Dapr NestJS integration instead :)

https://github.com/jeremylcarter/nest-dapr/tree/feature/actor

**Comment by cskiwi** [comment]
I made some progress that can be relevant, right now this is skipping the analog. So just having a NestJs + SSR and native angular. 

I think once I get it up and running it should be possible to port for analog. 
Repo: https://github.com/cskiwi/herring-gull-angular
Open issues (one is just nx vs vite related)
- https://github.com/nitedani/vite-plugin-angular/issues/10
- https://github.com/nitedani/vite-plugin-angular/issues/11

**Comment by cskiwi** [comment]
@brandonroberts, @jeremylcarter 

So I finally got some time to work on it, but i'm on a strange roadblock which I can't figure out.
I've made a new nx repo including 3 projects and more or less the latest package versions:
- A default analog project
- A Nestjs + Angular project
- A Nestjs + Analog project

So I'm currently having 2 issues,

## 1) Analog is not a function

If the `package.json` includes `"type": "module",` none of the analog projects build and they throw:

```
failed to load config from C:\Users\glenn\Documents\Code\cskiwi\analog-nestjs\apps\nestjs-analog\vite.config.ts

 >  NX   analog is not a function
```

## 2) vite_dev_server is not a function
If I remove the module type, then the non-analog project won't build with some ESM error, but the default analog project builds, the nestjs-analog project then gives me:

```

failed to load config from C:\Users\glenn\Documents\Code\cskiwi\analog-nestjs\apps\nestjs-analog\vite.config.ts

 >  NX   (0 , import_expose_vite_dev_server.default) is not a function
```

So here is where I'm currtly stuck.  If anyone wants to have a look at it, feel free to clone the following repo to get going: https://github.com/cskiwi/analog-nestjs
